File: api/app/services/safe_fetch.py
"""SSRF-safe HTTP fetcher.

Connects directly to validated IPs to prevent DNS rebinding.
Replaces trafilatura.fetch_url() with full redirect validation.
"""

import ipaddress
from urllib.parse import urljoin, urlparse

import urllib3

from app.services.url_validator import (
    UnsafeURLError,
    _resolve_with_timeout,
    is_safe_ip,
)

MAX_REDIRECTS = 5
REDIRECT_STATUSES = {301, 302, 303, 307, 308}
USER_AGENT = "Mozilla/5.0 (compatible; TechNewsCurator/1.0)"


def _resolve_and_validate(hostname: str) -> str:
    """Resolve hostname and validate all IPs. Return first safe IP.

    Raises:
        UnsafeURLError: All IPs are unsafe or hostname is an unsafe IP literal.
    """
    # IP literal
    try:
        ip = ipaddress.ip_address(hostname)
        if not is_safe_ip(ip):
            raise UnsafeURLError(f"IP address {ip} is not safe")
        return str(ip)
    except ValueError:
        pass

    # Domain: resolve all A/AAAA records
    addr_info = _resolve_with_timeout(hostname)

    safe_ip = None
    for info in addr_info:
        ip_str = info[4][0]
        ip = ipaddress.ip_address(ip_str)
        if not is_safe_ip(ip):
            raise UnsafeURLError(f"Hostname {hostname} resolves to unsafe IP {ip}")
        if safe_ip is None:
            safe_ip = ip_str

    if safe_ip is None:
        raise UnsafeURLError(f"No addresses found for {hostname}")

    return safe_ip


def _make_request(url: str) -> urllib3.HTTPResponse:
    """Parse URL, resolve + validate IP, connect directly to validated IP.

    DNS rebinding prevention: connect to resolved_ip with Host header
    set to original hostname. HTTPS uses server_hostname for SNI.
    """
    parsed = urlparse(url)
    scheme = parsed.scheme
    hostname = parsed.hostname
    port = parsed.port
    path = parsed.path or "/"
    if parsed.query:
        path = f"{path}?{parsed.query}"

    if scheme not in ("http", "https"):
        raise UnsafeURLError(f"Unsupported scheme: {scheme}")

    if not hostname:
        raise UnsafeURLError("No hostname in URL")

    resolved_ip = _resolve_and_validate(hostname)

    default_port = 443 if scheme == "https" else 80
    actual_port = port or default_port

    # Host header: include port only if non-default (RFC 7230 Section 5.4)
    host_header = hostname if port is None else f"{hostname}:{port}"

    if scheme == "https":
        pool = urllib3.HTTPSConnectionPool(
            host=resolved_ip,
            port=actual_port,
            server_hostname=hostname,
            timeout=urllib3.Timeout(connect=5, read=30),
            retries=False,
            maxsize=1,
            block=False,
        )
    else:
        pool = urllib3.HTTPConnectionPool(
            host=resolved_ip,
            port=actual_port,
            timeout=urllib3.Timeout(connect=5, read=30),
            retries=False,
            maxsize=1,
            block=False,
        )

    return pool.request(
        "GET",
        path,
        headers={
            "Host": host_header,
            "User-Agent": USER_AGENT,
        },
        redirect=False,
        assert_same_host=False,
    )


def safe_fetch(url: str) -> str | None:
    """Fetch URL content safely with SSRF validation on every redirect.

    Returns:
        HTML string, or None on fetch failure.

    Raises:
        UnsafeURLError: URL or redirect target is unsafe.
    """
    current_url = url
    for _ in range(MAX_REDIRECTS + 1):
        try:
            response = _make_request(current_url)
        except (urllib3.exceptions.HTTPError, OSError):
            return None

        if response.status in REDIRECT_STATUSES:
            location = response.headers.get("Location")
            if not location:
                return None
            current_url = urljoin(current_url, location)
            continue

        if response.status != 200:
            return None

        return response.data.decode("utf-8", errors="replace")

    return None  # Max redirects exceeded

File: api/src/services/safe-fetch.ts
import http from "node:http";
import https from "node:https";
import type tls from "node:tls";
import ipaddr from "ipaddr.js";
import { config } from "../config.js";
import type { AppLogger } from "../lib/logger.js";
import { rootLogger } from "../lib/logger.js";
import { buildBrowserHeaders } from "./browser-headers.js";
import { createCookieStore } from "./cookie-store.js";
import { decompressBody } from "./http-decompress.js";
import { buildTlsOptions } from "./tls-fingerprint.js";
import { UnsafeURLError, isSafeIp, resolveWithTimeout } from "./url-validator.js";

const MAX_REDIRECTS = 5;
const REDIRECT_STATUSES = new Set([301, 302, 303, 307, 308]);
const CONNECT_TIMEOUT = 5000;
const READ_TIMEOUT = 30000;
const MAX_BODY_SIZE = 10 * 1024 * 1024; // 10 MB

// Persist Cloudflare's __cf_bm cookie across ingests so its bot-score smoothing
// applies (see docs/DESIGN-ingest-fetch-hardening.md). Keyed by hostname only,
// so it does not affect SSRF IP validation.
const cookieStore = createCookieStore();

async function resolveAndValidate(hostname: string): Promise<string> {
  // IP literal
  const cleanHostname = hostname.replace(/^\[|\]$/g, "");
  if (ipaddr.isValid(cleanHostname)) {
    if (!isSafeIp(cleanHostname)) {
      throw new UnsafeURLError(`IP address ${cleanHostname} is not safe`);
    }
    return cleanHostname;
  }

  // Domain: resolve all records
  const addresses = await resolveWithTimeout(hostname);
  let safeIp: string | null = null;

  for (const addr of addresses) {
    if (!isSafeIp(addr)) {
      throw new UnsafeURLError(`Hostname ${hostname} resolves to unsafe IP ${addr}`);
    }
    if (safeIp === null) safeIp = addr;
  }

  if (safeIp === null) {
    throw new UnsafeURLError(`No addresses found for ${hostname}`);
  }

  return safeIp;
}

function buildRequestOptions(
  url: string,
  resolvedIp: string,
  cookieHeader: string,
): { options: http.RequestOptions | https.RequestOptions; isHttps: boolean } {
  const parsed = new URL(url);
  const isHttps = parsed.protocol === "https:";
  const defaultPort = isHttps ? 443 : 80;
  const port = parsed.port ? Number.parseInt(parsed.port, 10) : defaultPort;
  const path = parsed.pathname + parsed.search;

  // Host header: include port only if non-default
  const hostHeader = parsed.port ? `${parsed.hostname}:${parsed.port}` : parsed.hostname;

  const headers: http.OutgoingHttpHeaders = {
    ...buildBrowserHeaders(config.fetchUserAgent),
    Host: hostHeader,
  };
  if (cookieHeader) headers.Cookie = cookieHeader;

  const options: http.RequestOptions | https.RequestOptions = {
    hostname: resolvedIp,
    port,
    path,
    method: "GET",
    headers,
    timeout: CONNECT_TIMEOUT,
  };

  if (isHttps) {
    const tlsOptions = buildTlsOptions();
    // ALPNProtocols lives on tls.ConnectionOptions (honored at runtime by
    // https.request via tls.connect) but is absent from https.RequestOptions.
    const httpsOptions = options as https.RequestOptions & tls.ConnectionOptions;
    httpsOptions.servername = parsed.hostname;
    httpsOptions.rejectUnauthorized = true;
    httpsOptions.minVersion = tlsOptions.minVersion;
    httpsOptions.ALPNProtocols = tlsOptions.ALPNProtocols;
    httpsOptions.ciphers = tlsOptions.ciphers;
  }

  return { options, isHttps };
}

// Store Set-Cookie (e.g. __cf_bm) then decompress the body by Content-Encoding.
function finalizeResponse(
  res: http.IncomingMessage,
  cookieHost: string,
  chunks: Buffer[],
): { status: number; headers: http.IncomingHttpHeaders; body: string } {
  cookieStore.storeSetCookies(cookieHost, res.headers["set-cookie"], Date.now());
  const body = decompressBody(
    Buffer.concat(chunks),
    res.headers["content-encoding"],
    MAX_BODY_SIZE,
  ).toString("utf-8");
  return { status: res.statusCode ?? 0, headers: res.headers, body };
}

function makeRequest(
  url: string,
  resolvedIp: string,
): Promise<{ status: number; headers: http.IncomingHttpHeaders; body: string }> {
  return new Promise((resolve, reject) => {
    const cookieHost = new URL(url).hostname;
    const cookieHeader = cookieStore.getCookieHeader(cookieHost, Date.now());
    const { options, isHttps } = buildRequestOptions(url, resolvedIp, cookieHeader);

    const client = isHttps ? https : http;
    const req = client.request(options, (res) => {
      const chunks: Buffer[] = [];
      let totalBytes = 0;
      const readTimer = setTimeout(() => {
        req.destroy();
        reject(new Error("Read timeout"));
      }, READ_TIMEOUT);

      res.on("data", (chunk: Buffer) => {
        totalBytes += chunk.length;
        if (totalBytes > MAX_BODY_SIZE) {
          clearTimeout(readTimer);
          req.destroy();
          reject(new Error(`Response body too large (>${MAX_BODY_SIZE} bytes)`));
          return;
        }
        chunks.push(chunk);
      });
      res.on("end", () => {
        clearTimeout(readTimer);
        try {
          resolve(finalizeResponse(res, cookieHost, chunks));
        } catch (err) {
          reject(err);
        }
      });
      res.on("error", (err) => {
        clearTimeout(readTimer);
        reject(err);
      });
    });

    req.on("error", reject);
    req.on("timeout", () => {
      req.destroy();
      reject(new Error("Connect timeout"));
    });
    req.end();
  });
}

async function validateAndResolve(url: string): Promise<string> {
  const parsed = new URL(url);
  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    throw new UnsafeURLError(`Unsupported scheme: ${parsed.protocol.replace(":", "")}`);
  }
  if (!parsed.hostname) {
    throw new UnsafeURLError("No hostname in URL");
  }
  return resolveAndValidate(parsed.hostname);
}

export async function safeFetch(url: string, logger?: AppLogger): Promise<string | null> {
  const log =
    logger?.child({ service: "safe-fetch" }) ?? rootLogger.child({ service: "safe-fetch" });
  let currentUrl = url;

  for (let i = 0; i < MAX_REDIRECTS + 1; i++) {
    const resolvedIp = await validateAndResolve(currentUrl);

    let response: { status: number; headers: http.IncomingHttpHeaders; body: string };
    try {
      response = await makeRequest(currentUrl, resolvedIp);
    } catch (err) {
      log.warn(
        { url: currentUrl, error: err instanceof Error ? err.message : String(err) },
        "request failed",
      );
      return null;
    }

    if (REDIRECT_STATUSES.has(response.status)) {
      const location = response.headers.location;
      if (!location) {
        log.warn({ url: currentUrl, status: response.status }, "redirect without Location header");
        return null;
      }
      const nextUrl = new URL(location, currentUrl).toString();
      log.debug({ from: currentUrl, to: nextUrl, status: response.status }, "following redirect");
      currentUrl = nextUrl;
      continue;
    }

    if (response.status !== 200) {
      log.warn({ url: currentUrl, status: response.status }, "non-200 response");
      return null;
    }

    log.debug({ url: currentUrl, bodyLength: response.body.length }, "fetch successful");
    return response.body;
  }

  log.warn({ url, maxRedirects: MAX_REDIRECTS }, "max redirects exceeded");
  return null;
}
